# Analysis of video game sales, by Deborah Thomas.

<div style="background-color: rgb(255, 176, 155); padding: 10px; border-radius: 5px;">
    <h2>Introduction</h2>
</div>

#### Analysis of historic video game sales, from 1980-2016, from the (fictitional) online store called "Ice". This dataset includes sales from these three regions:
- North America
- Europe
- Japan
#### I will be analyzing sales, from these three regions, along with the following, in order to gain an understanding as to which video games will be successful:
- Video game platforms
- Year of release
- Video game genres
- Critics' scores
- Users' scores
- ESRB's ratings (Entertainment Software Rating Board)

<div style="background-color: rgb(255, 176, 155); padding: 10px; border-radius: 5px;">
    <h2>Import libraries, and read in the dataset</h2>
</div>

In [6]:
import pandas as pd
import numpy as np

import random

from IPython.display import Image

In [8]:
Image(url='../girl_videoGame_dog_ice.webp', width=300, height=300)

In [10]:
games = pd.read_csv('../games.csv')
display(games.head(5))

Name Platform  Year_of_Release         Genre  NA_sales  \
0                Wii Sports      Wii           2006.0        Sports     41.36   
1         Super Mario Bros.      NES           1985.0      Platform     29.08   
2            Mario Kart Wii      Wii           2008.0        Racing     15.68   
3         Wii Sports Resort      Wii           2009.0        Sports     15.61   
4  Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing     11.27   

   EU_sales  JP_sales  Other_sales  Critic_Score User_Score Rating  
0     28.96      3.77         8.45          76.0          8      E  
1      3.58      6.81         0.77           NaN        NaN    NaN  
2     12.76      3.79         3.29          82.0        8.3      E  
3     10.93      3.28         2.95          80.0          8      E  
4      8.89     10.22         1.00           NaN        NaN    NaN

<div style="background-color: rgb(255, 176, 155); padding: 10px; border-radius: 5px;">
    <h2>Basic summary of the data</h2>
</div>

In [13]:
print("This dataset has " + str(games.shape[1]) + " columns, and " + str(games.shape[0]) + " rows.")

This dataset has 11 columns, and 16715 rows.


In [15]:
games.describe()

Year_of_Release      NA_sales      EU_sales      JP_sales  \
count     16446.000000  16715.000000  16715.000000  16715.000000   
mean       2006.484616      0.263377      0.145060      0.077617   
std           5.877050      0.813604      0.503339      0.308853   
min        1980.000000      0.000000      0.000000      0.000000   
25%        2003.000000      0.000000      0.000000      0.000000   
50%        2007.000000      0.080000      0.020000      0.000000   
75%        2010.000000      0.240000      0.110000      0.040000   
max        2016.000000     41.360000     28.960000     10.220000   

        Other_sales  Critic_Score  
count  16715.000000   8137.000000  
mean       0.047342     68.967679  
std        0.186731     13.938165  
min        0.000000     13.000000  
25%        0.000000     60.000000  
50%        0.010000     71.000000  
75%        0.030000     79.000000  
max       10.570000     98.000000

#### North America had the most sales.

In [18]:
games.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16715 entries, 0 to 16714
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Name             16713 non-null  object 
 1   Platform         16715 non-null  object 
 2   Year_of_Release  16446 non-null  float64
 3   Genre            16713 non-null  object 
 4   NA_sales         16715 non-null  float64
 5   EU_sales         16715 non-null  float64
 6   JP_sales         16715 non-null  float64
 7   Other_sales      16715 non-null  float64
 8   Critic_Score     8137 non-null   float64
 9   User_Score       10014 non-null  object 
 10  Rating           9949 non-null   object 
dtypes: float64(6), object(5)
memory usage: 1.4+ MB


In [20]:
games.Critic_Score.max()

98.0

In [22]:
games.User_Score.value_counts()

User_Score
tbd    2424
7.8     324
8       290
8.2     282
8.3     254
       ... 
1.1       2
1.9       2
9.6       2
0         1
9.7       1
Name: count, Length: 96, dtype: int64

#### A quick study of the data shows:
- North America had the highest video game sales overall.
    
- Column names need to be lowercase.

- 'Year_of_Release' column should be renamed to 'year', to shorten the gap between the 'Platform' column.

- 'Year_of_Release', and 'Critic_Score' should not have decimals. These datatypes will need to change to int.
- 'User_Score' appears to only go to 10, so this can stay as a decimal. But, the datatype will need to change to float.
- The 'Rating' column will need to change to the 'category' datatype.

- These columns have NaN values and / or missing data: Name, 'Year_of_Release', 'Genre', 'Critic_Score', 'User_Score', 'Rating'.
- Luckily, there is no data missing from the sales columns: 'NA_sales', 'EU_sales', 'JP_sales', 'Other_sales'. But I will be changing the names to 'sales_na', 'sales_eu', 'sales_jp', 'sales_other', so the word 'sales' reads first.

<div style="background-color: rgb(255, 176, 155); padding: 10px; border-radius: 5px;">
    <h2>Clean the data</h2>
</div>

In [26]:
Image(url='../brooms.jpeg', width=300, height=300)

### Rename the column names to lowercase.

In [29]:
# New column names
new_columns = ['name', 'platform', 'year', 'genre', 'sales_na', 'sales_eu', 'sales_jp', 'sales_other', 'critic_score', 'user_score', 'rating']

# Assign the new column names to the DataFrame
games.columns = new_columns

#Display dataframe with new lowercase names
display(games.head(3))

name platform    year     genre  sales_na  sales_eu  sales_jp  \
0         Wii Sports      Wii  2006.0    Sports     41.36     28.96      3.77   
1  Super Mario Bros.      NES  1985.0  Platform     29.08      3.58      6.81   
2     Mario Kart Wii      Wii  2008.0    Racing     15.68     12.76      3.79   

   sales_other  critic_score user_score rating  
0         8.45          76.0          8      E  
1         0.77           NaN        NaN    NaN  
2         3.29          82.0        8.3      E

## Clean 'year' column

In [32]:
games.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16715 entries, 0 to 16714
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   name          16713 non-null  object 
 1   platform      16715 non-null  object 
 2   year          16446 non-null  float64
 3   genre         16713 non-null  object 
 4   sales_na      16715 non-null  float64
 5   sales_eu      16715 non-null  float64
 6   sales_jp      16715 non-null  float64
 7   sales_other   16715 non-null  float64
 8   critic_score  8137 non-null   float64
 9   user_score    10014 non-null  object 
 10  rating        9949 non-null   object 
dtypes: float64(6), object(5)
memory usage: 1.4+ MB


#### There are 269 NaN values in the 'year' column.

#### Some of the names of games have the year, explicitly listed as part of the title of the game. I will find just the titles that have 4 digits in their naming, and use that to impute the data for the NaN values in the 'year' column.

In [36]:
# Extract the 4-digit year from the 'name' column
year_extracted = games['name'].str.extract(r'(\d{4})')[0]

# Create a boolean mask for rows where 'name' contains a 4-digit year, and 'year' is NaN
mask = year_extracted.notna() & games['year'].isna()

# Filter the DataFrame using the mask
filtered_games = games[mask]

# Print the 'name', original 'year', and 'extracted_year' columns for verification
print("\nGames where 'name' contains a 4-digit year and 'year' column is NaN:")
filtered_games = filtered_games.assign(extracted_year=year_extracted[mask])
print(filtered_games[['name', 'year', 'extracted_year']])


Games where 'name' contains a 4-digit year and 'year' column is NaN:
                                    name  year extracted_year
183                      Madden NFL 2004   NaN           2004
377                     FIFA Soccer 2004   NaN           2004
475           wwe Smackdown vs. Raw 2006   NaN           2006
1650                 NASCAR Thunder 2003   NaN           2003
2572      PES 2009: Pro Evolution Soccer   NaN           2009
3486                     Madden NFL 2002   NaN           2002
4775                    NFL GameDay 2003   NaN           2003
5156                       NBA Live 2003   NaN           2003
5336                  Tomb Raider (2013)   NaN           2013
5655              All-Star Baseball 2005   NaN           2005
5889                       NBA Live 2003   NaN           2003
8918              All-Star Baseball 2005   NaN           2005
13195                Tour de France 2011   NaN           2011
13929                    Sega Rally 2006   NaN           2006


In [38]:
# Extract the 4-digit year from the 'name' column
year_extracted = games['name'].str.extract(r'(\d{4})')[0]

# Convert the extracted years to integers
year_extracted = pd.to_numeric(year_extracted, errors='coerce')

# This will print 'NaN' for the game 'name' that did not have a 4-digit number as part of their naming.
print(year_extracted.tail(20))

16695       NaN
16696       NaN
16697    3000.0
16698       NaN
16699       NaN
16700       NaN
16701       NaN
16702       NaN
16703       NaN
16704       NaN
16705       NaN
16706       NaN
16707       NaN
16708       NaN
16709    1000.0
16710       NaN
16711    2007.0
16712       NaN
16713       NaN
16714    2016.0
Name: 0, dtype: float64


#### I can see a 4-digit number higher than 2016, so I will have to only extract digits less than or equal to 2016.

In [41]:
# Create a boolean mask for rows where 'year' is NaN and extracted year <= 2016
mask = games['year'].isna() & year_extracted.notna() & (year_extracted <= 2016)

In [61]:
# Update the 'year' column for these masked rows with the extracted year
games.loc[mask, 'year'] = year_extracted[mask]

# Print the updated DataFrame
print("\nUpdated DataFrame with 'year' column filled from extracted 'year' (only for <= 2016):")
display(games.tail(30))


Updated DataFrame with 'year' column filled from extracted 'year' (only for <= 2016):


name platform    year  \
16685              Help Wanted: 50 Wacky Jobs (jp sales)      Wii  2008.0   
16686  Yattaman Wii: BikkuriDokkiri Machine de Mou Ra...      Wii  2008.0   
16687                              Neo Angelique Special      PSP  2008.0   
16688                      Outdoors Unleashed: Africa 3D      3DS  2011.0   
16689                                          Real Rode      PS2  2008.0   
16690                                     Pony Friends 2       PC  2009.0   
16691                       Dynasty Warriors: Eiketsuden      PS3  2016.0   
16692                  Metal Gear Solid V: Ground Zeroes       PC  2014.0   
16693                                  PGA European Tour      N64  2000.0   
16694                                     Carmageddon 64      N64  1999.0   
16695                                    Planet Monsters      GBA  2001.0   
16696                                             Breach       PC  2011.0   
16697                                   Bust-A-Move 3000       GC  2003.0   
16698                                   Mega Brain Boost       DS  2008.0   
16699                              The Longest 5 Minutes      PSV  2016.0   
16700                           Mezase!! Tsuri Master DS       DS  2009.0   
16701  Eiyuu Densetsu: Sora no Kiseki Material Collec...      PSP  2007.0   
16702                            STORM: Frontline Nation       PC  2011.0   
16703                                   Strawberry Nauts      PSV  2016.0   
16704                                           Plushees       DS  2008.0   
16705                                            15 Days       PC  2009.0   
16706                      Men in Black II: Alien Escape       GC  2003.0   
16707                                   Aiyoku no Eustia      PSV  2014.0   
16708                 Woody Woodpecker in Crazy Castle 5      GBA  2002.0   
16709   SCORE International Baja 1000: The Official Game      PS2  2008.0   
16710                      Samurai Warriors: Sanada Maru      PS3  2016.0   
16711                                   LMA Manager 2007     X360  2006.0   
16712                            Haitaka no Psychedelica      PSV  2016.0   
16713                                   Spirits & Spells      GBA  2003.0   
16714                                Winning Post 8 2016      PSV  2016.0   

              genre  sales_na  sales_eu  sales_jp  sales_other  critic_score  \
16685    Simulation      0.00      0.00      0.01          0.0           NaN   
16686        Racing      0.00      0.00      0.01          0.0           NaN   
16687     Adventure      0.00      0.00      0.01          0.0           NaN   
16688        Sports      0.01      0.00      0.00          0.0           NaN   
16689     Adventure      0.00      0.00      0.01          0.0           NaN   
16690    Simulation      0.00      0.01      0.00          0.0           NaN   
16691        Action      0.00      0.00      0.01          0.0           NaN   
16692        Action      0.00      0.01      0.00          0.0          80.0   
16693        Sports      0.01      0.00      0.00          0.0           NaN   
16694        Action      0.01      0.00      0.00          0.0           NaN   
16695        Action      0.01      0.00      0.00          0.0          67.0   
16696       Shooter      0.01      0.00      0.00          0.0          61.0   
16697        Puzzle      0.01      0.00      0.00          0.0          53.0   
16698        Puzzle      0.01      0.00      0.00          0.0          48.0   
16699        Action      0.00      0.00      0.01          0.0           NaN   
16700        Sports      0.00      0.00      0.01          0.0           NaN   
16701  Role-Playing      0.00      0.00      0.01          0.0           NaN   
16702      Strategy      0.00      0.01      0.00          0.0          60.0   
16703     Adventure      0.00      0.00      0.01          0.0           NaN   
16704    Simulation      0.01      0.00      0.00          0.0  

In [45]:
games.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16715 entries, 0 to 16714
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   name          16713 non-null  object 
 1   platform      16715 non-null  object 
 2   year          16463 non-null  float64
 3   genre         16713 non-null  object 
 4   sales_na      16715 non-null  float64
 5   sales_eu      16715 non-null  float64
 6   sales_jp      16715 non-null  float64
 7   sales_other   16715 non-null  float64
 8   critic_score  8137 non-null   float64
 9   user_score    10014 non-null  object 
 10  rating        9949 non-null   object 
dtypes: float64(6), object(5)
memory usage: 1.4+ MB


#### There are now only 252 NaN in the 'year' column. That method of imputing cleaned up 17 NaN values.

### I will find the most common year, for each game, then impute missing data, by filling in the 'year' with the most common year for each same-name game.

In [49]:
# Group by 'name' and 'year' and count occurrences
grouped_name_year = games.groupby(['name', 'year']).size().reset_index(name='count')

print("Grouped DataFrame with Counts:")
print(grouped_name_year.head(20))

Grouped DataFrame with Counts:
                                           name    year  count
0                                Beyblade Burst  2016.0      1
1                             Fire Emblem Fates  2015.0      1
2                          Frozen: Olaf's Quest  2013.0      2
3                    Haikyu!! Cross Team Match!  2016.0      1
4                             Tales of Xillia 2  2012.0      1
5                                   '98 Koshien  1998.0      1
6                    .hack//G.U. Vol.1//Rebirth  2006.0      1
7                  .hack//G.U. Vol.2//Reminisce  2006.0      1
8       .hack//G.U. Vol.2//Reminisce (jp sales)  2006.0      1
9                 .hack//G.U. Vol.3//Redemption  2007.0      1
10                      .hack//Infection Part 1  2002.0      1
11                                  .hack//Link  2010.0      1
12                       .hack//Mutation Part 2  2002.0      1
13                       .hack//Outbreak Part 3  2002.0      1
14  .hack//Quarantine Pa

In [51]:
# Determine the most common year for each game
#Apply a function to each 'year' group.
#Calculates the mode (most frequent value) of the 'year' values in the 'name' group.
#['0'] selects the first mode value if mode() returns multiple values. 
# Check if the mode calculation returns an empty result. If it does, it assigns np.nan instead. 
common_year = games.groupby('name')['year'].apply(lambda x: x.mode().iloc[0] if not x.mode().empty else None)

In [53]:
#Take the above info, and now fill in the missing values.  
#Create a dictionary from the common_year Series
common_year_dict = common_year.to_dict()

#Use this dictionary to fill in missing 'year' values
#If row is found in common_year_dict, it returns the most common year for that game.
games['year'] = games.apply(lambda row: common_year_dict.get(row['name'], row['year']) if pd.isna(row['year']) else row['year'], axis=1)

In [55]:
games.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16715 entries, 0 to 16714
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   name          16713 non-null  object 
 1   platform      16715 non-null  object 
 2   year          16576 non-null  float64
 3   genre         16713 non-null  object 
 4   sales_na      16715 non-null  float64
 5   sales_eu      16715 non-null  float64
 6   sales_jp      16715 non-null  float64
 7   sales_other   16715 non-null  float64
 8   critic_score  8137 non-null   float64
 9   user_score    10014 non-null  object 
 10  rating        9949 non-null   object 
dtypes: float64(6), object(5)
memory usage: 1.4+ MB


#### Now there are 139 NaN values in the 'year' column. That method of imputing cleaned up 113 NaN values.

## I need to figure out how to impute more data for the 'year' column.

#### The ESRB rating will give a clue as to what the year might be.
- The "T" (Teen) rating was introduced in the year 1994, so any game that has NaN in the 'year' column must be between 1994-2016. 
- The "E" (Everyone) rating was introduced in the year 1998, so any game that has NaN in the 'year' column must be between 1998-2016.
- The E10+ rating was not introduced until the year 2005, so any game that has NaN in the 'year' column must be between 2005-2016.

### I will iterate through each row of the dataframe, and look for 'rating' have value of "E10+", then impute a NaN value in 'year' with a random number from 2005-2016.  I will not do this with "E", nor "T", as those year ranges are too big, and there is less of a chance that I will impute with the correct year.

#### Also, the gaming platform would give as clue as to what year a game is.

In [162]:
games.platform.value_counts()

platform
PS2     2161
DS      2151
PS3     1331
Wii     1320
X360    1262
PSP     1209
PS      1197
PC       974
XB       824
GBA      822
GC       556
3DS      520
PSV      430
PS4      392
N64      319
XOne     247
SNES     239
SAT      173
WiiU     147
2600     133
NES       98
GB        98
DC        52
GEN       29
NG        12
SCD        6
WS         6
3DO        3
TG16       2
GG         1
PCFX       1
Name: count, dtype: int64

### Find the release year of each platform.

These are the years that the various gaming platforms were released. So, the NaN 'year' values, that match these platforms cannot be earlier than their platform's release date.

#### 1977
- 2600

#### 1983
- NES
- PC

#### 1967
- TG16

#### 1989
- GB

#### 1990
- GG
- SNES

#### 1991
- SCD

#### 1993
- 3DO

#### 1994
- NG
- PCFX
- PS
- SAT

#### 1996
- N64

#### 1999
- WS

#### 2000
- GC
- PS2

#### 2001
- GBA
- XB

#### 2002
- GEN

#### 2004
- PSP

#### 2005
- DS

#### 2005.
- X360

#### 2006
- PS3
- Wii

#### 2010
- 3DS

#### 2011
- DC

#### 2012
- WiiU

- PSV

#### 2013
- PS4
- XOne

### Make a dictionary with those platform values and platform year of release.

In [163]:
platform_release_years = {
    1967: ['PC'],
    1977: ['2600'],
    1983: ['NES'],
    1989: ['TG16', 'GB'],
    1990: ['GG'],
    1991: ['SNES', 'SCD'],
    1993: ['3DO'],
    1994: ['NG', 'PCFX', 'PS', 'SAT'],
    1996: ['N64'],
    1999: ['WS'],
    2000: ['GC', 'PS2'],
    2001: ['GBA', 'XB'],
    2002: ['GEN'],
    2004: ['PSP'],
    2005: ['DS', 'X360'],
    2006: ['PS3', 'Wii'],
    2010: ['3DS'],
    2011: ['DC'],
    2012: ['WiiU', 'PSV'],
    2013: ['PS4', 'XOne']
}

In [164]:
# Initialize a dictionary to store counts
nan_platform_counts = {}

In [165]:
# Iterate through each list of platforms in the dictionary
for platforms in platform_release_years.values():
    for platform in platforms:
        # Filter the games DataFrame for the current platform and where year is NaN
        filtered_games = games[(games['platform'] == platform) & (games['year'].isna())]
        
        # Count occurrences
        count = filtered_games['platform'].value_counts().to_dict()
        
        # Update counts in the nan_platform_counts dictionary
        if platform in nan_platform_counts:
            nan_platform_counts[platform] += count.get(platform, 0)
        else:
            nan_platform_counts[platform] = count.get(platform, 0)
# Convert the counts dictionary to a DataFrame
nan_platform_value_counts = pd.DataFrame(list(nan_platform_counts.items()), columns=['platform', 'count'])

# Sort the DataFrame by the 'count' column in descending order
nan_platform_value_counts = nan_platform_value_counts.sort_values(by='count', ascending=False)

# Display the result
print("Value counts of games where 'year' is NaN:")
print(nan_platform_value_counts)

Value counts of games where 'year' is NaN:
   platform  count
21       DS     22
24      Wii     19
16      PS2     17
1      2600     16
23      PS3     11
22     X360     10
17      GBA      9
11       PS      7
20      PSP      6
18       XB      6
25      3DS      5
0        PC      4
15       GC      4
13      N64      1
28      PSV      1
4        GB      1
14       WS      0
29      PS4      0
27     WiiU      0
26       DC      0
2       NES      0
3      TG16      0
5        GG      0
6      SNES      0
7       SCD      0
19      GEN      0
8       3DO      0
9        NG      0
10     PCFX      0
12      SAT      0
30     XOne      0


### For years greater or equal to 2004, I will take the year value, from the platform_release_years dictionary, then fill in the NaN values, in the 'year' column of the dataframe, with a random number that is between the year the platform was released to the year 2016, but only if the 'platform' matches the value in the platform_release_years dictionary.

In [171]:
# Iterate through the dictionary and update the 'year' column for keys >= 2004
for year, platforms in platform_release_years.items():
    if year >= 2004:
        # Filter the DataFrame where 'platform' matches and 'year' is NaN
        mask = games['platform'].isin(platforms) & games['year'].isna()
        # Update the 'year' column for the matching rows with a random year between 'year' and 2016
        games.loc[mask, 'year'] = games.loc[mask].apply(
            lambda row: random.randint(year, 2016), axis=1
        )
# Display the updated DataFrame
print("Updated DataFrame with 'year' column filled for platforms released after 2004:")
display(games)

Updated DataFrame with 'year' column filled for platforms released after 2004:


name platform  year         genre  sales_na  \
0                         Wii Sports      Wii  2006        Sports     41.36   
1                  Super Mario Bros.      NES  1985      Platform     29.08   
2                     Mario Kart Wii      Wii  2008        Racing     15.68   
3                  Wii Sports Resort      Wii  2009        Sports     15.61   
4           Pokemon Red/Pokemon Blue       GB  1996  Role-Playing     11.27   
...                              ...      ...   ...           ...       ...   
16710  Samurai Warriors: Sanada Maru      PS3  2016        Action      0.00   
16711               LMA Manager 2007     X360  2006        Sports      0.00   
16712        Haitaka no Psychedelica      PSV  2016     Adventure      0.00   
16713               Spirits & Spells      GBA  2003      Platform      0.01   
16714            Winning Post 8 2016      PSV  2016    Simulation      0.00   

       sales_eu  sales_jp  sales_other  critic_score user_score rating  
0         28.96      3.77         8.45          76.0          8      E  
1          3.58      6.81         0.77           NaN        NaN    NaN  
2         12.76      3.79         3.29          82.0        8.3      E  
3         10.93      3.28         2.95          80.0          8      E  
4          8.89     10.22         1.00           NaN        NaN    NaN  
...         ...       ...          ...           ...        ...    ...  
16710      0.00      0.01         0.00           NaN        NaN    NaN  
16711      0.01      0.00         0.00           NaN        NaN    NaN  
16712      0.00      0.01         0.00           NaN        NaN    NaN  
16713      0.00      0.00         0.00           NaN        NaN    NaN  
16714      0.00      0.01         0.00           NaN        NaN    NaN  

[16715 rows x 11 columns]

In [172]:
games.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16715 entries, 0 to 16714
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   name          16713 non-null  object 
 1   platform      16715 non-null  object 
 2   year          16715 non-null  int64  
 3   genre         16713 non-null  object 
 4   sales_na      16715 non-null  float64
 5   sales_eu      16715 non-null  float64
 6   sales_jp      16715 non-null  float64
 7   sales_other   16715 non-null  float64
 8   critic_score  8137 non-null   float64
 9   user_score    10014 non-null  object 
 10  rating        9949 non-null   object 
dtypes: float64(5), int64(1), object(5)
memory usage: 1.4+ MB


#### Now there are only 65 NaN values in the 'year' column. Imputing by this method cleaned up 74 NaN.

#### Fill the remaining 74 NaN values, in the 'year' column, to '0'. Change datatype to int.

In [173]:
#Fill the remaining 146 NaN values with '0' before changing type from float to int. 
#Get rid of decimal in year column.
games['year'] = games['year'].fillna(0).astype(int)

# Verify the changes
print("\nDataFrame after cleaning 'year' column:")
display(games)


DataFrame after cleaning 'year' column:


name platform  year         genre  sales_na  \
0                         Wii Sports      Wii  2006        Sports     41.36   
1                  Super Mario Bros.      NES  1985      Platform     29.08   
2                     Mario Kart Wii      Wii  2008        Racing     15.68   
3                  Wii Sports Resort      Wii  2009        Sports     15.61   
4           Pokemon Red/Pokemon Blue       GB  1996  Role-Playing     11.27   
...                              ...      ...   ...           ...       ...   
16710  Samurai Warriors: Sanada Maru      PS3  2016        Action      0.00   
16711               LMA Manager 2007     X360  2006        Sports      0.00   
16712        Haitaka no Psychedelica      PSV  2016     Adventure      0.00   
16713               Spirits & Spells      GBA  2003      Platform      0.01   
16714            Winning Post 8 2016      PSV  2016    Simulation      0.00   

       sales_eu  sales_jp  sales_other  critic_score user_score rating  
0         28.96      3.77         8.45          76.0          8      E  
1          3.58      6.81         0.77           NaN        NaN    NaN  
2         12.76      3.79         3.29          82.0        8.3      E  
3         10.93      3.28         2.95          80.0          8      E  
4          8.89     10.22         1.00           NaN        NaN    NaN  
...         ...       ...          ...           ...        ...    ...  
16710      0.00      0.01         0.00           NaN        NaN    NaN  
16711      0.01      0.00         0.00           NaN        NaN    NaN  
16712      0.00      0.01         0.00           NaN        NaN    NaN  
16713      0.00      0.00         0.00           NaN        NaN    NaN  
16714      0.00      0.01         0.00           NaN        NaN    NaN  

[16715 rows x 11 columns]

In [174]:
games.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16715 entries, 0 to 16714
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   name          16713 non-null  object 
 1   platform      16715 non-null  object 
 2   year          16715 non-null  int64  
 3   genre         16713 non-null  object 
 4   sales_na      16715 non-null  float64
 5   sales_eu      16715 non-null  float64
 6   sales_jp      16715 non-null  float64
 7   sales_other   16715 non-null  float64
 8   critic_score  8137 non-null   float64
 9   user_score    10014 non-null  object 
 10  rating        9949 non-null   object 
dtypes: float64(5), int64(1), object(5)
memory usage: 1.4+ MB


### 'year' column is clean now.

## Clean 'critic_score' column

#### There are currently 8578 NaN values, which is about half of the rows in the 'critic_score' column.

### Impute the data for 'critic_score'

In [ ]:
games.critic_score.value_counts()

In [ ]:
# Group by 'genre' and calculate the median critic_score for each group
genre_medians_critic = games.groupby('genre')['critic_score'].transform('median')

# Fill NaN values in 'critic_score' column with the Critic's median score for that genre
games['critic_score'] = games['critic_score'].fillna(genre_medians_critic)

In [ ]:
games.info()

#### A few of the columns have only 2 rows with missing values. Could it be that there are 2 rows that have missing values in many rows? If so, those two rows should be dropped from the dataframe.

In [ ]:
# Filter rows where 'critic_score' is NaN
nan_critic_score_rows = games[games['critic_score'].isna()]

# Display the filtered rows
print("Rows with NaN values in 'critic_score':")
display(nan_critic_score_rows)

#### Yes, those 2 rows have multiple columns with NaN values. These 2 rows will be dropped from the dataframe.

In [ ]:
# Drop the 2 rows where 'critic_score' is NaN
games = games.dropna(subset=['critic_score'])

# Verify the changes by displaying the modified DataFrame
print("\nDataFrame after dropping rows with NaN values in 'critic_score':")
display(games.head(5))

In [ ]:
games['critic_score'] = games['critic_score'].astype(int)

## Clean 'user_score' column

### Get rid of NaN from 'user_score' column.

In [ ]:
games.info()

In [ ]:
#### There are 10014 rows with non-null in the 'user_score' column. 

In [ ]:
# Counting the number of NaN values in the 'user_score' column
nan_count_user_score = games['user_score'].isna().sum()

print(f"\nNumber of NaN values in 'user_score': {nan_count_user_score}")

#### There are 16713 rows. There are 10014 non-null values in the 'user_score' column. There are 9123 NaN values in the 'user_score' column.  10014 - 9123 = 891. There are 891 unaccounted non-null values that are not NaN. What are they?

In [ ]:
# Identify non-numeric/ non-null values in 'user_score' column.
# Convert to numeric and identify non-numeric values. The 'errors='coerce' ensures that any non-numeric entries are converted to NaN.
games['user_score_numeric'] = pd.to_numeric(games['user_score'], errors='coerce')
invalid_values_user_score = games[games['user_score_numeric'].isna() & games['user_score'].notna()]

print("\nNon-numeric non-null values in 'user_score':")
print(invalid_values_user_score[['name', 'user_score']])

In [ ]:
# Inspect unique values and their counts
user_score_value_counts = games['user_score'].value_counts(dropna=False)

print("\nUnique values and their counts in 'user_score':")
print(user_score_value_counts)

#### Since the user_score only goes to 10, I will impute the median score, for each genre. I will leave the decimals.

In [ ]:
games['user_score'] = games['user_score'].replace('tbd', np.nan)

### Change 'user_score' datatype to float.

In [ ]:
# Change 'user_score' from float to int type.
games['user_score'] = games['user_score'].astype(float)

### Impute data for 'user_score'.

In [ ]:
# Group by 'genre' and calculate the median user_score for each group
genre_medians_user = games.groupby('genre')['user_score'].transform('median')

# Fill NaN values in 'user_score' column with the User's median score for that genre
games['user_score'] = games['user_score'].fillna(genre_medians_user)

In [ ]:
games.info()

## Clean 'rating' column.

### Get rid of NaN values in the 'rating' column by imputing data.

In [ ]:
games.genre.value_counts()

In [ ]:
games.rating.value_counts()

#### Find games that contain certain words in their title, that suggest that the game is not an all-ages game.

In [ ]:
# List of keywords implying a mature rating
keywords = ['evil', 'war', 'kill', 'devil', 'gun', 'battle', 'fatal', 'hell', 'psychedelic']

# Create a mask to find rows where 'name' contains any of the keywords
mask_rating = games['name'].str.contains('|'.join(keywords), case=False, na=False)

# Display rows where the mask is True (i.e., 'name' contains keywords)
mature_games = games[mask_rating]

print("Games implying mature rating based on keywords in 'name':")
display(mature_games.tail(40))

In [ ]:
#From this list, find only the games with a NaN 'rating' value.
mask_nan_rating = mask_rating & games['rating'].isna()

#Change the 'rating' to 'M' (mature), or 'T' (teen)
#mask_nan_rating is boolean.
games.loc[mask_nan_rating, 'rating'] = [random.choice(['M', 'T']) for _ in range(mask_nan_rating.sum())]

#### Check that the 'rating', of words on the keywords list,  were randomly changed to 'M' or 'T'.

In [ ]:
mask_check = games['name'].str.contains('|'.join(keywords), case=False, na=False)

filtered_names = games.loc[mask_check, ['name', 'rating']]
print(filtered_names)

In [ ]:
# Filter rows where genre is 'Platform'
platform_games = games[games['genre'] == 'Platform']

# Print the 'title', 'rating', and 'genre' columns of the filtered rows
print(platform_games[['name', 'rating', 'genre']])

#### If one "Super Mario" game has an "E" rating, then I should be able to fill in Nan values, of other "Super Mario" games with an "E" rating.

In [ ]:
# Identify rows where 'name' contains 'Super Mario' and 'rating' is NaN, then fill with 'E'
games.loc[games['name'].str.contains('Super Mario', case=False) & games['rating'].isna(), 'rating'] = 'E'

In [ ]:
#After cleaning ...
# Filter rows where genre is 'Platform'
platform_games = games[games['genre'] == 'Platform']

# Print the 'title', 'rating', and 'genre' columns of the filtered rows
print(platform_games[['name', 'rating', 'genre']].head(5))

### If that worked for "Super Mario", then it will work to get rid of NaN values for "Pokemon", too.

In [ ]:
# Replace 'Pokémon' with 'Pokemon' in the 'name' column
games['name'] = games['name'].str.replace('Pokémon', 'Pokemon')

In [ ]:
# Filter rows where 'name' contains 'Pokemon'
name_contains_pokemon = games[games['name'].str.contains('Pokemon', case=False, na=False)]

# Display the filtered DataFrame with specified columns only
print(name_contains_pokemon[['name', 'rating', 'genre']])

In [ ]:
# After cleaning
# Filter rows where 'name' contains 'Pokemon'
name_contains_pokemon = games[games['name'].str.contains('Pokemon', case=False, na=False)]

# Display the filtered DataFrame with specified columns only
print(name_contains_pokemon[['name', 'rating', 'genre']])

In [ ]:
# Identify rows where 'name' contains 'Pokemon' and 'rating' is NaN, then fill with 'E'
games.loc[games['name'].str.contains('Pokemon', case=False) & games['rating'].isna(), 'rating'] = 'E'

In [ ]:
# After cleaning
# Filter rows where 'name' contains 'Pokemon'
name_contains_pokemon = games[games['name'].str.contains('Pokemon', case=False, na=False)]

# Display the filtered DataFrame with specified columns only
print(name_contains_pokemon[['name', 'rating', 'genre']])

### Do the same method with game names that contain "Zelda".

In [ ]:
# Filter rows where 'name' contains 'Zelda'
name_contains_zelda = games[games['name'].str.contains('Zelda', case=False, na=False)]

# Display the filtered DataFrame
print(name_contains_zelda[['name', 'rating', 'genre']].tail(30))

In [ ]:
# Identify rows where 'name' contains 'Zelda' and 'rating' is NaN, then fill with 'E'
games.loc[games['name'].str.contains('Zelda', case=False) & games['rating'].isna(), 'rating'] = 'E'

In [ ]:
#After cleaning ...
# Filter rows where 'name' contains 'Zelda'
name_contains_zelda = games[games['name'].str.contains('Zelda', case=False, na=False)]

# Display the filtered DataFrame
print(name_contains_zelda[['name', 'rating', 'genre']].tail(30))

### Impute data on 'rating' column using bfill and ffil.

In [ ]:
# Group by 'name' and impute NaNs in 'rating' column with available ratings for the same name
games['rating'] = games.groupby('name')['rating'].transform(lambda x: x.ffill().bfill())

# Verify the updates
display(games.sample(20))

In [ ]:
games.name.value_counts()

In [ ]:
# Filter rows where genre is 'Fighting'
platform_games = games[games['genre'] == 'Fighting']

# Print the 'title', 'rating', and 'genre' columns of the filtered rows
print(platform_games[['name', 'rating', 'genre']])

In [ ]:
#Before cleaning ...
# Filter rows where 'name' contains 'Super Smash Bros.'
name_contains_super_smash_bros = games[games['name'].str.contains('Super Smash Bros.', case=False, na=False)]

# Display the filtered DataFrame
print(name_contains_super_smash_bros[['name', 'rating', 'genre']].tail(30))

In [ ]:
# Identify rows where 'name' contains 'Super Smash Bros.' and 'rating' is NaN, then fill with 'E'
games.loc[games['name'].str.contains('Super Smash Bros.', case=False) & games['rating'].isna(), 'rating'] = 'T'

In [ ]:
#After cleaning ...
# Filter rows where 'name' contains 'Super Smash Bros.'
name_contains_super_smash_bros = games[games['name'].str.contains('Super Smash Bros.', case=False, na=False)]

# Display the filtered DataFrame
print(name_contains_super_smash_bros[['name', 'rating', 'genre']].tail(30))

In [ ]:
# Filter rows where genre is 'Puzzle'
platform_games_strategy = games[games['genre'] == 'Puzzle']

# Print the 'title', 'rating', and 'genre' columns of the filtered rows
print(platform_games_strategy[['name', 'rating', 'genre']].head(25))

In [ ]:
tetris_games = games[games['name'].str.contains('Tetris', case=False, na=False)]
display(tetris_games)

In [ ]:
#Change all 'name' that contain 'Tetris' to a 'rating' of 'E'.
games.loc[games['name'].str.contains('Tetris', case=False, na=False), 'rating'] = 'E'

In [ ]:
games.info()

In [ ]:
#Replace the remainder of NaN values with "None".
games['rating'] = games['rating'].fillna('None')

In [ ]:
games['rating'] = games['rating'].astype('category')

In [ ]:
games.info()

### Remove temporary columns that were used for cleaning.

In [ ]:
games.drop(columns=['user_score_numeric'], inplace=True)

<div style="background-color: rgb(255, 176, 155); padding: 10px; border-radius: 5px;">
    <h2>Data is clean now.</h2>
</div>

<br>
<h2><b>Why were there missing values?</b></h2>
<br>

<b>'rating'</b>
<br>

- The ESRB rating system was not created until 1994. This dataset goes back to 1980. So, for the first 15 years of the dataset, I wouldn't expect to find a value in the 'rating' column. Since the dataset goes until 2016, this explains why almost half of the values for the 'rating' column were not there.

- In 2005, ESRB introduced a new rating, E10+, so I wouldn't expect to find any of those ratings before the year 2005.

- Another explanation for missing values in the 'rating' column is that whoever created the dataset just failed to fill in that field.

<br>
<b>'critic_score'</b>
<br>
<br>
- There are so many games produced each year. The less popular games probably did not get scored by critics.
<br>
<br>

<b>'user_score'</b>
<br>
<br>
- Again, the less popular games probably did not get scored by users.

- Also, how did the users score the game before the invention of the internet? Did they snail mail their vote in? It would be unlikely that many people would take the time to do that. Even after the invention of the internet, it took many years before every household actually had the internet. This could also explain missing 'user_score'.

<h2><b>Explanation of Cleaning:</h2></b>
<br>
<br>
<b>'year'</b>

- I used groupby, to group the name of same-named games together. Then I used lambda to to get the year mode(the most frequent value), of each name of game, then apply it to each same-named game that had NaN. 

- If the mode calculation returned an empty result, it was assigned np.nan instead. 

- I used fillna to fill the remainder of the NaN values to '0'.

- I got rid of the decimal in the year column, and changed the datatype to int64, as years do not have decimals. I converted 'year' to Int64, instead of int, as Int64 handles NaN values gracefully without raising an error.

<br>
<br>
<b>'critic_score'</b>

- I grouped by genre, and filled the NaN values, in the 'critic_score' column with the median values of each genre's 'critic_score' values.

- After doing this, there were only 2 rows, from the 'critic_score' column that had missing values. I printed out those 2 rows, and almost all the values were missing for the other rows, too, so I dropped those 2 rows from the dataframe.

- I changed the datatype to 'int', as decimals seemed excessive, considering that the scale goes to 100.

<br>
<br>
<b>'user_score'</b>

- I changed to datatype float, leaving the decimals in, because the 'user_score' column only goes to 10, so those decimals give an extra added degree of rating.

- I identified non-numeric/ non-null values in 'user_score' column, using pd.to_numeric, and used 'errors='coerce' to ensure that any non-numeric entries were converted to NaN.

- Then I used 'replace' to change 'tbd' to np.nan

- I used groupby to group by 'genre' and 'user_score' to find the median 'user_score' for each genre.

- I took those values, and used fillna to fill in the missing 'user_score' values with the median 'user_score' of that game's genre.

<br>
<br>
<b>'rating'</b>

- I created a keywords list that contained words that might imply that a game is not an all ages game.

- I then created a mask, to find rows where 'name' contains any of the words on the keywords list.

- I took that list, and found all games from that list that had a NaN value in the 'rating' column.

- I used random to replace 'NaN, in all of those games' 'rating' column, with either 'M' ( for "Mature" ), or 'T' ( for "Teen ).

- I queried the dataframe, by different genres, to see if there were any games that I recognized, and I studied those games to see if any of the other games, with similiar names, had a value for 'rating'. If they did, I used str.contains and .isna to find games of that name, that had NaN for their 'rating' value, and I changed the 'rating' to the same rating as a similar named game. I did this for "Zelda", "Pokemon", "Tetris", "Super Smash Bros.", and "Super Mario" games.

- I then grouped by 'name' and 'rating, and used ffill and bfill to ffill, to grab from previous game, and grab from the next game on the list, from within the group.

- I went back up, a few cells, and used replace, to change the name of 'Pokémon' to  'Pokemon', because the 'Pokémon'with the apostropher above the 'e' did not show up in searches when I searched for "Pokemon", and so the change in 'rating' did not take effect. So, now all "Pokemon" show up when I search for them.

- The remainder of missing values were filled with 'None', using fillna.

- I changed the datatype to 'category', as this may be helpful when making graphs.

<br>
<br>
<b>Removed temp columns</b>

- I used drop to remove the temporary column called 'user_score_numeric'


<div style="background-color: rgb(255, 176, 155); padding: 10px; border-radius: 5px;">
    <h2>Analyze the data.</h2>
</div>

## Calculate the total sales ( the sum of each region ), for each game.

In [ ]:
#Initialize total_sales
games['total_sales'] = 0.0

In [ ]:
# Calculate the total sales for each game by summing 'sales_na', 'sales_eu', and 'sales_jp'
games['total_sales'] = games['sales_na'] + games['sales_eu'] + games['sales_jp']

# Display the DataFrame to verify the new column
print("DataFrame with total sales calculated for each game:")
display(games)

In [ ]:
# Group by 'year' and count the number of games released each year
release_by_year = games.groupby('year').size().reset_index(name='num_games')

In [ ]:
# Sort by 'year' 
release_by_year = release_by_year.sort_values(by='year')

In [ ]:
display(release_by_year)

#### The most games, per year, were released between the years 2006-2011.

In [ ]:
# Sort by 'year', from highest to lowest
release_by_year = release_by_year.sort_values(by='num_games', ascending=False)
display(release_by_year)

#### After 1993, there starts to be a lot more games released per year.